fdbcli/TenantCommands.actor.cpp

```diff
@@ -368,28 +412,42 @@ ACTOR Future<bool> tenantListCommand(Reference<IDatabase> db, std::vector<String
 			tr->setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
 			state ClusterType clusterType = wait(TenantAPI::getClusterType(tr));
 			state std::vector<TenantName> tenantNames;
+			// State filters only apply to calls from the management cluster
+			// Tenant group filters can apply to management, data, and standalone clusters
 			if (clusterType == ClusterType::METACLUSTER_MANAGEMENT) {
-				if (filters.empty()) {
+				std::vector<std::pair<TenantName, metacluster::MetaclusterTenantMapEntry>> tenants = wait(
+				    metacluster::listTenantMetadata(db, beginTenant, endTenant, limit, offset, filters, tenantGroup));
+				if (useJson) {
+					tenantListOutputJson(tenants);
+					return true;
+				} else {
+					for (const auto& [tenantName, _] : tenants) {
+						tenantNames.push_back(tenantName);
+					}
+				}
+			} else {
+				if (tenantGroup.present()) {
+					// For expediency: does not use special key space
+					// TODO: add special key support
 					std::vector<std::pair<TenantName, int64_t>> tenants =
-					    wait(metacluster::listTenants(db, beginTenant, endTenant, limit, offset));
-					for (auto tenant : tenants) {
-						tenantNames.push_back(tenant.first);
+					    wait(TenantAPI::listTenantGroupTenants(db, tenantGroup.get(), beginTenant, endTenant, limit));
+					if (useJson) {
+						tenantListOutputJson(tenants);
+						return true;
+					} else {
+						for (const auto& [tenantName, tenantId] : tenants) {
+							tenantNames.push_back(tenantName);
+						}
 					}
 				} else {
-					std::vector<std::pair<TenantName, metacluster::MetaclusterTenantMapEntry>> tenants =
-					    wait(metacluster::listTenantMetadata(db, beginTenant, endTenant, limit, offset, filters));
+					// Hold the reference to the standalone's memory
+					state ThreadFuture<RangeResult> kvsFuture =
+					    tr->getRange(firstGreaterOrEqual(beginTenantKey), firstGreaterOrEqual(endTenantKey), limit);
+					RangeResult tenants = wait(safeThreadFutureToFuture(kvsFuture));
 					for (auto tenant : tenants) {
-						tenantNames.push_back(tenant.first);
+						tenantNames.push_back(tenant.key.removePrefix(tenantMapSpecialKeyRange.begin));
```

Comment:
This part doesn't work if you specify `JSON` currently. To get it to work, we'll need to parse the value (a JSON string) and extract its ID.

I don't know if it makes things simpler, but you could probably refactor each of these three blocks (management cluster, non-management w/ groups, non-management w/o groups) to all generate a map from name->ID (or the reverse). Then you could have one check to generate JSON or not at the end, and you would only need one function to print the JSON instead of two.